pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -414,18 +417,22 @@ cdef class {{name}}HashTable(HashTable):
             int64_t[:] labels
             Py_ssize_t idx, count = count_prior
             int ret = 0
-            {{dtype}}_t val
+            {{dtype}}_t val, na_value
             khiter_t k
             {{name}}VectorData *ud
+            bint use_na_value
 
         labels = np.empty(n, dtype=np.int64)
         ud = uniques.data
+        na_value = self.na_value
+        use_na_value = self.use_na_value
 
         with nogil:
             for i in range(n):
                 val = values[i]
 
-                if check_null and {{null_condition}}:
+                if ((check_null and {{null_condition}}) or
```

Comment:
Related to above comment, is it not possible to do something like 

```python
if (check_null and (val != val or val == na_value)):
```

To still cover all of the scenarios (borrowed the above logic from `groupby_helper`)